feat(robot-server): add instrument firmware update and monitoring capability

sanni-t opened the pull request:
Closes RCORE-482, RCORE-483

# Overview

Adds endpoints to the robot-server to facilitate updating instruments' firmware. These endpoints will be primarily used to update instruments that get attached *after* a robot has finished auto-updating firmware of the instruments attached during a robot update.

The work for this feature started in a different branch whose code was reviewed [here](https://github.com/Opentrons/opentrons/pull/12209) if you want to take a look. Had to create a separate PR for untangling the multiple merge conflicts and for git tree readability.

# Test Plan

Test updating OT3 instruments and fetching the update status 
NOTE- you will need to temporarily modify the robot environment manually to make it so that it thinks there's an update available.-

1. Send a `POST` request to `/instruments/updates` endpoint with the mount specified in the body as
```
{"mount": <mount name>}
```
2. Fetch update status by sending `GET` request to `/instruments/updates/{:update_id}` where `update_id` is a unique ID generated by the `POST` update request.
3. Test that the following conditions make the update to fail with an appropriate error-
  - No instrument attached on the mount
  - Initiating an update while that mount is already updating
  - Initiating an update when there's no new firmware version available
  - Trying to fetch update status using a non-existent update ID
4. Check how long does it take for an update request to respond (useful for testing the timeout we can use for waiting for the update status)

Anything else you can think about.

# Changelog

- added the above two endpoints to update & fetch instruments.
- added an `UpdateProgressMonitor` class to keep track of the update processes that were started

# Review requests

- Please go through the code and check if it makes sense 
- Check that I have covered all test cases
- Check that inline comments/ todos make sense
- Test it on a robot! Bonus points if you could take screenshots of the responses so we can use them while wiring up the UI components.

# Risk assessment

Low for the code in edge. But could be detrimental to an instrument itself if an update process fails mid-way.

sanni-t committed:
initial firmware update endpoint

Also adds todos for tests and errors

sanni-t committed:
restructured endpoints and added an update progress monitor

sanni-t committed:
background the update task, add firmware versions fields

wait before fetching progress, UpdateProgressMonitor should be singleton, other misc improvements

refactored usage of fail_after

sfoster1 commented on robot-server/robot_server/instruments/instrument_models.py:
```diff
@@ -57,6 +80,17 @@ class _GenericInstrument(GenericModel, Generic[InstrumentModelT, InstrumentDataT
     instrumentModel: InstrumentModelT = Field(..., description="Instrument model.")
     # TODO (spp, 2023-01-06): add firmware version field
     serialNumber: str = Field(..., description="Instrument hardware serial number.")
+    currentFirmwareVersion: Optional[int] = Field(
```
These are optionals to support the OT-2, right? I'd love to not have it work like that by building a second endpoint wrapper or something if it's a flex, but failing that maybe we can bundle them up in a subdict/submodel that can be optional in its entirety, so it doesn't look like you can have this be partially filled?

b-cooper replied:
I'd be fine with a subdict too, that said, if there is a way for OT2's to return reasonable information in these fields as well, that could work too.

sfoster1 commented on robot-server/robot_server/instruments/instrument_models.py:
```diff
@@ -57,6 +80,17 @@ class _GenericInstrument(GenericModel, Generic[InstrumentModelT, InstrumentDataT
     instrumentModel: InstrumentModelT = Field(..., description="Instrument model.")
     # TODO (spp, 2023-01-06): add firmware version field
```
done!

sfoster1 commented on robot-server/robot_server/instruments/router.py:
```diff
@@ -26,10 +31,83 @@
+    return update_progress_monitor
+
+
+class InstrumentNotFound(ErrorDetails):
```
this feels really general, is this really the first time we've had an error like this? could we put it somewhere more common?

sfoster1 commented on robot-server/robot_server/instruments/router.py:
```diff
@@ -26,10 +31,83 @@
+    title: str = "An update is already in progress."
+
+
+class NotSupportedOnOT2(ErrorDetails):
```
this one too, actually - do we do a similar check anywhere else?

sanni-t replied:
This is the first router that has features that are different for OT2 & OT3 so I doubt there's a similar check elsewhere but I'll confirm

sfoster1 commented on robot-server/robot_server/instruments/router.py:
```diff
@@ -26,10 +31,83 @@
+    title: str = "Cannot update OT2 instruments' firmware."
+
+
+class InvalidUpdateId(ErrorDetails):
```
this feels kind of overspecific to me. it's only raised in the endpoint where you get an update by id, so we already know that it's an invalid update id, that's the only kind of id there is there. a more general invalid id might be more reusable, I'd think?

sfoster1 commented on robot-server/robot_server/instruments/router.py:
```diff
@@ -122,3 +206,114 @@ async def get_attached_instruments(
+        ).as_error(status.HTTP_404_NOT_FOUND)
+
+    if ot3_hardware.get_firmware_update_progress().get(ot3_mount):
+        raise UpdateInProgress(
```
do you think it's worth providing a redirect to the status for the update?

sanni-t replied:
So instead of raising an error, just bypass calling `update_instrument_firmware` and respond with the in-process update progress status?

sfoster1 commented on robot-server/robot_server/instruments/update_progress_monitor.py:
```diff
@@ -0,0 +1,118 @@
+        """Update the status for the given id."""
+        try:
+            ot3_hardware = ensure_ot3_hardware(hardware_api=self._hardware)
+        except HardwareNotSupportedError as e:
```
do we need this? all it does is cut off the traceback

sfoster1 reviewed (changes requested):
I have one thing I'm wondering about and one suggestion that I think would make this code a little more robust.

Wondering: There's a lot of custom exceptions that are used in basically the same kind of place (like an `UpdateIdNotFound(KeyError)`, for instance) - is this the server code style? it leads to a ton of nested-across-the-exception-stack catch and reraise, which is sort of noisy. 

There's also these custom error responses; some of them feel like they could be more generalized, which in addition to code style benefits will really help the effort to enumerate and categorize error responses to make life easier on the client. Those custom error responses have to have their HTTP status codes specified at the build site, which sort of feels duplicative if we're specifying everything else about the error; and finally, even though we have all these custom exception classes that have 1:1 correspondences with those error classes, we still have to manually catch those exceptions and then build a response - couldn't middleware help here?

That stuff is all more minor or out of scope though. The thing that I think should change is the way we're doing the update status handling, and it can change in a way that makes it easier to deal with and understand.

The core problem that's dealt with in `update_progress_monitor` is that we're running the update in an asyncio task; we're then polling for a snapshot of the update state; and that update state can change out from under us or even disappear, and we have to use heuristics to decide whether that means it's done or it fails. That's the only way to do it with a polling API like this, which is annoying. 

The thing is, we don't have to have that API, and in fact we already have more than that API. We have an async generator API that we can use instead, which we can capture and normalize in a similar way to how we do it at lower levels:

- Change `OT3API.update_instrument_firmware` to return `AsyncIterator[Set[UpdateStatus]]` as `update_firmware` does, and replace the line in that function that just dumps the update status into the debug log with `yield update_status`
- Make `UpdateProgressMonitor` just `Update` or something similar
- Give it an internal `asyncio.Queue[UpdateStatus]` 
- Give it an asyncio `Event` and an async `run` function, the core of which is `for status in self._hardware.update_firmware(...): self._event.set(); await self._queue.put(status)`; also give it an async `started` function that just does `await self._event.wait()`
- Give it a `_latest_status` member and an async `get_current_status` function; the core of the function is 
   ```
   while True: 
        try: 
             self._latest_status = self._queue.get_nowait() 
        except asyncio.QueueEmpty: 
             break
        return transform_status_to_appropriate_type(self._latest_status)
   ```

- When you start an update, create this object, stash it in the by-update-id store, and drop its `run` into an asyncio task. Replace busywaiting on the status to see when the update starts with `await update.started()` (appropriate `asyncio.wait()` call for timeouts), which will only clear after the first status update
- When you're asking for status, do it by calling `get_status()`, which should now always return the latest status info from the update, and massage it into the right shape

You can even take your `run` function and wrap that for-in with a try/except, and submit a special method for "error in update" to the queue that can be pulled by the status getter; you can submit a special "update successful" message to the queue if the for-in exits without an error.

This now does no polling, no busywaiting, won't miss messages, and doesn't need heuristics. I think it's worth the change.

sfoster1 committed:
factor some errors into common locations

sfoster1 committed:
lil changes

sfoster1 committed:
these changes are less little

sfoster1 committed:
the tests all pass

vegano1 reviewed (approved):
Ok this makes sense to me

sfoster1 committed:
catch and log exceptions properly

sfoster1 committed:
add a lovely script

sfoster1 committed:
refactor(api): ThreadManager handles async gens

Async generator functions are different from normal async functions, in the same
way that generator functions are different from normal functions. When called,
they return a special async generator object. This is a problem in
ThreadManager land, because those async generator objects need to be
interacted with via `await anext(gen)` in the same loop that created
them.

So, this commit adds a new case in thread manager for "what if you call
an async generator function in the hardware controller". In that case,
what happens is
- We return a proxy async generator function, which is defined in (and
therefore runs in) the calling loop
- We generate a future that is a coroutine running in the hardware
controller loop that captures a queue, and pulls items out of the async
generator in that loop and puts them in a queue
- Our proxy pulls stuff out of the queue and yields this

This is of course extra frustrating because all queue interaction must
take place in the same loop, which must be the loop the queue was
created in. That means that for the queue that moves generator output
from the hardware thread/loop to the calling thread/loop, it has to be
- Created in the hardware thread/loop (via run_coroutine_threadsafe and
a wrapper)
- Inserted into in the hardware thread/loop (fine, this happens in the
coroutine that drains the generator)
- Gotten from in the hardware thread/loop, so every await queue.get()
needs to happen in a run_coroutine_threadsafe()

It all works! But it's pretty annoying.

sfoster1 commented:
Superceded by #12504
